esteid/actions.py

```diff
@@ -1,86 +1,178 @@
+import os
+from tempfile import NamedTemporaryFile
+
 import binascii
+import logging
+import typing
+
+from .bdoc2.container import BDoc2File
+from .bdoc2.exceptions import BDoc2Error
+from .bdoc2.signature_verifier import verify_cryptography
+from .bdoc2.utils import prepare_signature, finalize_signature
+from .bdoc2.xmlsig import XmlSignature
+from .digidocservice.service import DigiDocError
+from .session import open_container, get_esteid_session, update_esteid_session, delete_esteid_session
+
+if typing.TYPE_CHECKING:
+    from .generic import GenericDigitalSignViewMixin
+
 
-from .digidocservice.service import DigiDocError, DigiDocNotOk
+logger = logging.getLogger(__name__)
 
 
 class BaseAction(object):
     @classmethod
-    def do_action(cls, view, action_kwargs):
+    def do_action(cls, view: "GenericDigitalSignViewMixin", params):
         raise NotImplementedError
 
 
 class NoAction(object):
     @classmethod
-    def do_action(cls, view, action_kwargs):
+    def do_action(cls, view, params):
         return {'success': True}
 
 
 class IdCardPrepareAction(BaseAction):
     @classmethod
-    def do_action(cls, view, action_kwargs):
-        service = view.get_service()
+    def do_action(cls, view: "GenericDigitalSignViewMixin", params: dict = None, *, certificate: str = None,
+                  container_path: str = None):
+        """
+        The old API is to pass a dict of params (previously confusingly named `action_kwargs`),
+        the keyword args are added here for clarity as to what the method accepts
 
-        if not view.get_files():
-            return {
-                'success': False,
-                'code': 'MIN_1_FILE',
-            }
+        :param view:
+        :param params:
+        :param certificate: HEX-encoded certificate from the ID card
+        :return:
+        """
+        request = view.request
 
-        # Create signed document
-        service.create_signed_document()
+        if not certificate:
+            certificate = params['certificate']
 
-        # add all files
-        for file in view.get_files():
-            service.add_datafile(file.file_name, file.mimetype, service.HASHCODE, file.size, file.content)
+        if container_path is None:
```

Comment:
Does this `container_path` come from the session (context: security and prevention of opening unwanted files)?

Reply:
`container_path` comes from wherever the user of this action deems necessary. It is just the path to an external container that the signature is about to be added.
On a side note, by _from the session_ do you mean `request.session`? If so, there is no security risk because `request.session` is an entity entirely controlled by the server code, and client can not affect it in any way. Django stores user auth data in sessions so it can't be more secure.

Reply:
Thanks for the clarification, just wanted to be sure